servers/physics_3d/shape_3d_sw.cpp

```diff
@@ -1640,8 +1654,177 @@ Vector3 HeightMapShape3DSW::get_support(const Vector3 &p_normal) const {
 	return get_aabb().get_support(p_normal);
 }
 
+void HeightMapShape3DSW::_face_cull_segment(_SegmentCullParams &p_params) const {
+	Vector3 res;
+	Vector3 normal;
+	if (p_params.face->intersect_segment(p_params.from, p_params.to, res, normal)) {
+		real_t d = p_params.dir.dot(res) - p_params.dir.dot(p_params.from);
+		if ((d > 0) && (d < p_params.min_d)) {
+			p_params.min_d = d;
+			p_params.result = res;
+			p_params.normal = p_params.face->normal;
+			p_params.normal = normal;
+			p_params.collisions++;
+		}
+	}
+}
+
+void HeightMapShape3DSW::_cell_cull_segment(_SegmentCullParams &p_params, int p_x, int p_z) const {
+	// First triangle.
+	_get_point(p_x, p_z, p_params.face->vertex[0]);
+	_get_point(p_x + 1, p_z, p_params.face->vertex[1]);
+	_get_point(p_x, p_z + 1, p_params.face->vertex[2]);
+	p_params.face->normal = Plane(p_params.face->vertex[0], p_params.face->vertex[1], p_params.face->vertex[2]).normal;
+	_face_cull_segment(p_params);
+
+	// Second triangle.
+	_get_point(p_x + 1, p_z, p_params.face->vertex[0]);
+	_get_point(p_x + 1, p_z + 1, p_params.face->vertex[1]);
+	p_params.face->normal = Plane(p_params.face->vertex[0], p_params.face->vertex[1], p_params.face->vertex[2]).normal;
+	_face_cull_segment(p_params);
+}
+
 bool HeightMapShape3DSW::intersect_segment(const Vector3 &p_begin, const Vector3 &p_end, Vector3 &r_point, Vector3 &r_normal) const {
-	return false;
+	if (heights.is_empty()) {
+		return false;
+	}
+
+	Vector3 local_begin = p_begin + local_origin;
```

Comment:
What if the ray begins outside of the heightmap? If it starts outside but intersects the AABB it should ideally start on the box hit

Reply:
That's what this code does, just before the loop:
```
// Start inside the grid.
int x_start = CLAMP(x, 0, width - 2);
int z_start = CLAMP(z, 0, depth - 2);

// Adjust initial cross values.
cross_x += delta_x * x_step * (x_start - x);
cross_z += delta_z * z_step * (z_start - z);

x = x_start;
z = z_start;
```

But I agree it might be better to just reduce the ray itself according to the AABB, to replace this and the check to stop when outside the grid within the loop.

Reply:
Given that it's already skipping the beginning and end of the ray and the current version seems to work fine, I actually haven't added the extra AABB check to change the ray.